kazuki-toolkit/src/main/kotlin/com/anaplan/engineering/kazuki/toolkit/iso8601/Dtg.kt

```diff
@@ -164,39 +164,42 @@ object DtgUtilities {
 
     val monthsBetweenDtgs: (Dtg, Dtg) -> nat = function(
         command = { earlierDtg, laterDtg ->
-            MonthsPerYear * yearsBetweenDtgs(earlierDtg, laterDtg) +
-                    (if (laterDtg.date.month < earlierDtg.date.month) 12uL else 0uL) +
-                    if (laterDtg.date.day >= earlierDtg.date.day) {
-                        laterDtg.date.month - earlierDtg.date.month
-                    } else {
-                        laterDtg.date.month - earlierDtg.date.month - 1uL
-                    }
-        },
-        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg }
-    )
+            val earlierDate = earlierDtg.date
+            val laterDate = laterDtg.date
 
-    val yearsBetweenDtgs: (Dtg, Dtg) -> nat = function(
-        command = { earlierDtg, laterDtg ->
+            val baseYears = laterDate.year - earlierDate.year
+            val baseMonths = ((baseYears * 12U) + laterDate.month) - earlierDate.month
 
-            val durationInYearUpToEarlierDtg =
-                earlierDtg.properties.durationSinceFirstDtg.functions.subtractDuration(
-                    Duration.durationFromFirstYearUpToStartOfYear(
-                        earlierDtg.date.year
-                    )
-                )
-            val durationInYearUpToLaterDtg =
-                laterDtg.properties.durationSinceFirstDtg.functions.subtractDuration(
-                    Duration.durationFromFirstYearUpToStartOfYear(
-                        laterDtg.date.year
-                    )
-                )
+            // The above calculation is off by one if laterDate is earlier on in its month than earlierDate
+            // (for example, 1 March is only one month after 2 January)
+            val partialMonth =
+                laterDate.day < earlierDate.day || laterDate.day == earlierDate.day && laterDtg.time < earlierDtg.time
 
-            if (durationInYearUpToEarlierDtg <= durationInYearUpToLaterDtg) {
-                laterDtg.date.year - earlierDtg.date.year
+            if (partialMonth) {
+                baseMonths - 1U
             } else {
-                laterDtg.date.year - earlierDtg.date.year - 1uL
+                baseMonths
             }
         },
-        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg }
+        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg },
+        post = { earlierDtg, laterDtg, result ->
+            val x = result.toLong()
+
+            earlierDtg.functions.addMonths(x) <= laterDtg &&
+                    earlierDtg.functions.addMonths(x + 1) > laterDtg
+        }
+    )
+
+    val yearsBetweenDtgs: (Dtg, Dtg) -> nat = function(
+        command = { earlierDtg, laterDtg ->
+            monthsBetweenDtgs(earlierDtg, laterDtg) / 12U
+        },
+        pre = { earlierDtg, laterDtg -> earlierDtg <= laterDtg },
+        post = { earlierDtg, laterDtg, result ->
+            val x = result.toLong()
```

Comment:
:pick: We'd want `toInteger()` here to use 'spec types'

A better fix would be to bring `addMonths/subtractMonths` in line with all the other functions and take `nat` rather than having these functions both take an int